From 5089a48ada309dc6669556835d7c5f1f5dddcf11 Mon Sep 17 00:00:00 2001
From: Ganesh Babu Jothiram <quic_gjothira@quicinc.com>
Date: Tue, 4 Apr 2023 17:47:17 +0530
Subject: [PATCH 1/2] ath12k: sawf: add uplink configuration parameters in
 service class paramters.

extend support for uplink configuration parameter in service class creation.

Signed-off-by: Ganesh Babu Jothiram <quic_gjothira@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/sawf.c   |  6 +++
 drivers/net/wireless/ath/ath12k/sawf.h   | 18 +++++++++
 drivers/net/wireless/ath/ath12k/vendor.c | 51 +++++++++++++++++++++++-
 drivers/net/wireless/ath/ath12k/vendor.h |  6 +++
 4 files changed, 80 insertions(+), 1 deletion(-)

diff --git a/drivers/net/wireless/ath/ath12k/sawf.c b/drivers/net/wireless/ath/ath12k/sawf.c
index a2f5acf..bfb30dd 100644
--- a/drivers/net/wireless/ath/ath12k/sawf.c
+++ b/drivers/net/wireless/ath/ath12k/sawf.c
@@ -108,6 +108,12 @@ void ath12k_update_svc_class(struct ath12k_sawf_svc_params *sawf_params)
 	new_param->tid = sawf_params->tid;
 	new_param->msdu_rate_loss = sawf_params->msdu_rate_loss;
 	new_param->configured = sawf_params->configured;
+	new_param->ul_service_interval = sawf_params->ul_service_interval;
+	new_param->ul_burst_size = sawf_params->ul_burst_size;
+	new_param->ul_min_tput = sawf_params->ul_min_tput;
+	new_param->ul_max_latency = sawf_params->ul_max_latency;
+	new_param->ul_ofdma_disable = sawf_params->ul_ofdma_disable;
+	new_param->ul_mu_mimo_disable = sawf_params->ul_mu_mimo_disable;
 }
 
 bool ath12k_validate_sawf_param(struct ath12k_sawf_svc_params *params)
diff --git a/drivers/net/wireless/ath/ath12k/sawf.h b/drivers/net/wireless/ath/ath12k/sawf.h
index 7824357..0eaa573 100644
--- a/drivers/net/wireless/ath/ath12k/sawf.h
+++ b/drivers/net/wireless/ath/ath12k/sawf.h
@@ -106,6 +106,12 @@ enum SAWF_SVC_PARAM_DEFAULTS {
 	SAWF_SVC_PARAM_DEFAULT_PRIORITY       = 0,
 	SAWF_SVC_PARAM_DEFAULT_TID            = 0xffffffff,
 	SAWF_SVC_PARAM_DEFAULT_MSDU_LOSS_RATE = 0,
+	SAWF_SVC_PARAM_DEFAULT_UL_BURST_SIZE  = 0,
+	SAWF_SVC_PARAM_DEFAULT_UL_MIN_TPUT    = 0,
+	SAWF_SVC_PARAM_DEFAULT_UL_MAX_LATENCY = 0xffffffff,
+	SAWF_SVC_PARAM_DEFAULT_UL_SVC_INTERVAL = 0xffffffff,
+	SAWF_SVC_PARAM_DEFAULT_UL_OFDMA_DISABLE = 0,
+	SAWF_SVC_PARAM_DEFAULT_UL_MU_MIMO_DISABLE = 0,
 };
 
 #define DELAY_BOUND_ULTRA_LOW 10
@@ -131,6 +137,12 @@ enum SAWF_SVC_PARAM_DEFAULTS {
  * @priority: Priority
  * @tid: TID
  * @msdu_rate_loss: MSDU loss rate in parts per million
+ * @ul_service_interval: Uplink service interval
+ * @ul_burst_size: Uplink Burst Size
+ * @ul_min_tput: Uplink min_throughput
+ * @ul_max_latency: Uplink max latency
+ * @ul_ofdma_disable: Disable ofdma
+ * @ul_mu_mimo_disable: Disale MU MIMO
  * @configured: indicating if the serivice class is configured.
  */
 
@@ -146,6 +158,12 @@ struct ath12k_sawf_svc_params {
 	u32 priority;
 	u32 tid;
 	u32 msdu_rate_loss;
+	u32 ul_service_interval;
+	u32 ul_burst_size;
+	u32 ul_min_tput;
+	u32 ul_max_latency;
+	bool ul_ofdma_disable;
+	bool ul_mu_mimo_disable;
 	bool configured;
 };
 
diff --git a/drivers/net/wireless/ath/ath12k/vendor.c b/drivers/net/wireless/ath/ath12k/vendor.c
index e64307a..eb7df8d 100644
--- a/drivers/net/wireless/ath/ath12k/vendor.c
+++ b/drivers/net/wireless/ath/ath12k/vendor.c
@@ -23,6 +23,12 @@ ath12k_vendor_sawf_config_policy[QCA_WLAN_VENDOR_SAWF_ATTR_CONFIG_MAX + 1] = {
 	[QCN_WLAN_VENDOR_ATTR_SAWF_SVC_PRIO] = {.type = NLA_U32},
 	[QCN_WLAN_VENDOR_ATTR_SAWF_SVC_TID] = {.type = NLA_U32},
 	[QCN_WLAN_VENDOR_ATTR_SAWF_SVC_MSDU_RATE_LOSS] = {.type = NLA_U32},
+	[QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_SVC_INTERVAL] = {.type = NLA_U32},
+	[QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_MIN_TPUT] = {.type = NLA_U32},
+	[QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_MAX_LATENCY] = {.type = NLA_U32},
+	[QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_BURST_SIZE] = {.type = NLA_U32},
+	[QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_OFDMA_DISABLE] = {.type = NLA_U8},
+	[QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_MU_MIMO_DISABLE] = {.type = NLA_U8},
 };
 
 static const struct nla_policy
@@ -47,6 +53,12 @@ static void ath12k_set_default_values(struct ath12k_sawf_svc_params *param)
 	param->priority = SAWF_SVC_PARAM_DEFAULT_PRIORITY;
 	param->tid = SAWF_SVC_PARAM_DEFAULT_TID;
 	param->msdu_rate_loss = SAWF_SVC_PARAM_DEFAULT_MSDU_LOSS_RATE;
+	param->ul_burst_size = SAWF_SVC_PARAM_DEFAULT_UL_BURST_SIZE;
+	param->ul_min_tput = SAWF_SVC_PARAM_DEFAULT_UL_MIN_TPUT;
+	param->ul_max_latency = SAWF_SVC_PARAM_DEFAULT_UL_MAX_LATENCY;
+	param->ul_service_interval = SAWF_SVC_PARAM_DEFAULT_UL_SVC_INTERVAL;
+	param->ul_ofdma_disable = SAWF_SVC_PARAM_DEFAULT_UL_OFDMA_DISABLE;
+	param->ul_mu_mimo_disable = SAWF_SVC_PARAM_DEFAULT_UL_MU_MIMO_DISABLE;
 }
 
 static int ath12k_vendor_set_sawf_config(struct wiphy *wiphy,
@@ -101,6 +113,24 @@ static int ath12k_vendor_set_sawf_config(struct wiphy *wiphy,
 		sawf_param.tid = nla_get_u32(tb[QCN_WLAN_VENDOR_ATTR_SAWF_SVC_TID]);
 	if (tb[QCN_WLAN_VENDOR_ATTR_SAWF_SVC_MSDU_RATE_LOSS])
 		sawf_param.msdu_rate_loss = nla_get_u32(tb[QCN_WLAN_VENDOR_ATTR_SAWF_SVC_MSDU_RATE_LOSS]);
+	if (tb[QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_SVC_INTERVAL])
+		sawf_param.ul_service_interval =
+			nla_get_u32(tb[QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_SVC_INTERVAL]);
+	if (tb[QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_BURST_SIZE])
+		sawf_param.ul_burst_size =
+			nla_get_u32(tb[QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_BURST_SIZE]);
+	if (tb[QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_MIN_TPUT])
+		sawf_param.ul_min_tput =
+			nla_get_u32(tb[QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_MIN_TPUT]);
+	if (tb[QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_MAX_LATENCY])
+		sawf_param.ul_max_latency =
+			nla_get_u32(tb[QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_MAX_LATENCY]);
+	if (tb[QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_OFDMA_DISABLE])
+		sawf_param.ul_ofdma_disable =
+			nla_get_u8(tb[QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_OFDMA_DISABLE]);
+	if (tb[QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_MU_MIMO_DISABLE])
+		sawf_param.ul_mu_mimo_disable =
+			nla_get_u8(tb[QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_MU_MIMO_DISABLE]);
 
 	ret = ath12k_create_send_svc_params(&sawf_param);
 
@@ -235,7 +265,26 @@ static int ath12k_vendor_view_sawf_config(struct wiphy *wihpy,
 				svc_class->tid) ||
 		    nla_put_u32(msg,
 				QCN_WLAN_VENDOR_ATTR_SAWF_SVC_MSDU_RATE_LOSS,
-				svc_class->msdu_rate_loss))
+				svc_class->msdu_rate_loss) ||
+		    nla_put_u32(msg,
+				QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_SVC_INTERVAL,
+				svc_class->ul_service_interval) ||
+		    nla_put_u32(msg,
+				QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_MIN_TPUT,
+				svc_class->ul_min_tput) ||
+		    nla_put_u32(msg,
+				QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_MAX_LATENCY,
+				svc_class->ul_max_latency) ||
+		    nla_put_u32(msg,
+				QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_BURST_SIZE,
+				svc_class->ul_burst_size) ||
+		    nla_put_u8(msg,
+				QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_OFDMA_DISABLE,
+				svc_class->ul_ofdma_disable) ||
+		    nla_put_u8(msg,
+				QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_MU_MIMO_DISABLE,
+				svc_class->ul_mu_mimo_disable))
+
 			goto nla_put_failure;
 
 		nest_end_length = nla_nest_end(msg, sawf_svc_class);
diff --git a/drivers/net/wireless/ath/ath12k/vendor.h b/drivers/net/wireless/ath/ath12k/vendor.h
index 855e72f..6a793d7 100644
--- a/drivers/net/wireless/ath/ath12k/vendor.h
+++ b/drivers/net/wireless/ath/ath12k/vendor.h
@@ -32,6 +32,12 @@ enum qca_wlan_vendor_sawf_attr_config {
 	QCN_WLAN_VENDOR_ATTR_SAWF_SVC_PRIO,
 	QCN_WLAN_VENDOR_ATTR_SAWF_SVC_TID,
 	QCN_WLAN_VENDOR_ATTR_SAWF_SVC_MSDU_RATE_LOSS,
+	QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_SVC_INTERVAL,
+	QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_MIN_TPUT,
+	QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_MAX_LATENCY,
+	QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_BURST_SIZE,
+	QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_OFDMA_DISABLE,
+	QCN_WLAN_VENDOR_ATTR_SAWF_SVC_UL_MU_MIMO_DISABLE,
 
 	/* keep last */
 	QCA_WLAN_VENDOR_SAWF_ATTR_CONFIG_AFTER_LAST,
-- 
2.17.1

